src/json/serialization.cr

```diff
@@ -124,6 +127,36 @@ module JSON
       # so it overloads well with other possible initializes
 
       def self.new(pull : ::JSON::PullParser)
+        {% verbatim do %}
+          {% begin %}
+            {% discriminators = @type.instance_vars.select { |ivar| ivar.annotation(::JSON::Discriminator) } %}
+            {% if !discriminators.empty? %}
+              {% if discriminators.size > 1 %}
+                {% raise "Multiple JSON::Discriminator annotations found for type #{@type}, instance variables #{discriminators.map { |var| "@#{var}" }.join(", ").id}" %}
+              {% end %}
+              {% discriminator = discriminators[0] %}
+              {% mapping = discriminator.annotation(::JSON::Discriminator)[0] %}
+              {% unless mapping.is_a?(HashLiteral) || mapping.is_a?(NamedTupleLiteral) %}
+                {% raise "Argument to JSON::Discriminator must be a HashLiteral ot NamedTupleLiteral, not #{mapping}" %}
+              {% end %}
+              any = JSON::Any.new(pull)
```

Comment:
The best way to do this is to do
```cr
data = IO::Memory.new
JSON.build(data) do |builder|
  builder.start_object
  pull.read_object do |key|
    if key == {{discriminator.id.stringify}}
      discriminator_value = pull.read_string
    end
    builder.field(key) { pull.read_raw(builder) }
  end
  builder.end_object    
end
data.rewind

...

```

Reply:
In theory we could optimize further by building a `IO::Union` (basically extract the `IO` concatenation feature from  `ARGF`). Then break the above loop as soon as the discriminator is found and pass a `JSON::PullParser.new(IO::Union.new(data, pull.io))`  to the child.

Reply:
@RX14 I think that's the best way to do it because we retain the original JSON string. My way of using `JSON::Any` is a bit broken with, for example, large numbers.

Reply:
Sure, don't feel strong about the naming.

Reply:
I suppose the discriminator field is likely the first one 80% of the time. In that case, optimization would be pretty simple, because we could essentially just delegate the pull parser to the subtype (there only needs to be some way to skip/repeat the start object event, but that's doable). This completely avoids the entire parse-build loop.